oauth2/handler.go

```diff
@@ -1154,3 +1207,175 @@ func (h *Handler) logOrAudit(err error, r *http.Request) {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("the format is not supported")))
+		return
+	}
+	if request.Proof == nil {
+		// Handle priming request
+		nonceLifespan := h.r.Config().GetVerifiableCredentialsNonceLifespan(ctx)
+		nonceExpiresIn := time.Now().Add(nonceLifespan).UTC()
+		nonce, err := h.r.OAuth2Storage().NewNonce(ctx, accessToken, nonceExpiresIn)
+		if err != nil {
+			h.r.Writer().WriteError(w, r, errorsx.WithStack(err))
+			return
+		}
+		w.WriteHeader(http.StatusBadRequest)
+		h.r.Writer().Write(w, r, &VerifiableCredentialPrimingResponse{
+			RFC6749ErrorJson: fosite.RFC6749ErrorJson{
+				Name:        "missing_proof",
+				Description: "Could not issue a verifiable credential because the proof is missing in the request.",
+			},
+			Format:          "jwt_vc",
+			Nonce:           nonce,
+			NonceExperiesIn: int64(nonceLifespan.Seconds()),
+		})
+		return
+	}
+	if request.Proof.ProofType != "jwt" {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("the proof type is not supported")))
+		return
+	}
+
+	header, _, ok := strings.Cut(request.Proof.JWT, ".")
+	if !ok {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("invalid JWT in proof")))
+		return
+	}
+
+	rawHeader, err := base64.RawURLEncoding.DecodeString(header)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("invalid JWT header in proof")))
+		return
+	}
+	jwk := gjson.GetBytes(rawHeader, "jwk").String()
+	proofJWK, err := josex.LoadJSONWebKey([]byte(jwk), true)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("invalid JWK in JWT header in proof")))
+		return
+	}
+
+	token, err := jwt.Parse(request.Proof.JWT, func(token *jwt.Token) (any, error) {
+		return proofJWK, nil
+	})
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("bad signature for JWT in proof")))
+		return
+	}
+
+	nonce, ok := token.Claims["nonce"].(string)
+	if !ok {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("nonce claim is missing")))
+		return
+	}
+
+	if err = h.r.OAuth2Storage().IsNonceValid(ctx, accessToken, nonce); err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("nonce is invalid")))
+		return
+	}
+
+	var response VerifiableCredentialResponse
+	response.Format = "jwt_vc_json"
+
+	proofJWKJSON, err := json.Marshal(proofJWK)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(err))
+		return
+	}
+
+	session.Claims.Add("vc", map[string]any{
+		"@context": []string{"https://www.w3.org/2018/credentials/v1"},
+		"type":     []string{"VerifiableCredential", "UserInfoCredential"},
+		"credentialSubject": map[string]any{
+			"id": fmt.Sprintf("did:jwk:%s", base64.RawURLEncoding.EncodeToString(proofJWKJSON)),
```

Comment:
Is this the correct encoding?

Reply:
https://github.com/quartzjer/did-jwk/blob/main/spec.md